compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -655,33 +655,40 @@ impl<'hir> LoweringContext<'_, 'hir> {
 
             hir::ExprKind::Closure(c)
         };
-        let parent_has_track_caller = self
-            .attrs
-            .values()
-            .find(|attrs| attrs.into_iter().find(|attr| attr.has_name(sym::track_caller)).is_some())
-            .is_some();
+
         let unstable_span =
             self.mark_span_with_reason(DesugaringKind::Async, span, self.allow_gen_future.clone());
 
-        let hir_id = if parent_has_track_caller {
-            let generator_hir_id = self.lower_node_id(closure_node_id);
-            self.lower_attrs(
-                generator_hir_id,
-                &[Attribute {
-                    kind: AttrKind::Normal(ptr::P(NormalAttr {
-                        item: AttrItem {
-                            path: Path::from_ident(Ident::new(sym::track_caller, span)),
-                            args: AttrArgs::Empty,
+        let hir_id = if self.tcx.features().closure_track_caller {
+            let parent_has_track_caller = self
+                .attrs
+                .values()
```

Comment:
This is from your previous PR, but why is this correct? Shouldn't we be checking the attributes of only the enclosing item, not all enclosing items recursively?

Reply:
if i remember correctly, what i was trying to do was go one level up in the span hierarchy, because the attribute was attached to the parent span, not the span that is passed into this function. but maybe there is a better way to achieve that to avoid the nested find

Reply:
I tried printing out `self.attrs` here (`self` being the `LoweringContext`) and I get this (for the `panic-track-caller.rs` test)
```
INFO rustc_ast_lowering::expr self.attrs ⚠️  are SortedMap { data: [] }
INFO rustc_ast_lowering::expr self.attrs ⚠️  are SortedMap { data: [] }
INFO rustc_ast_lowering::expr self.attrs ⚠️  are SortedMap { data: [(0, [Attribute { kind: Normal(NormalAttr { item: AttrItem { path: Path { span: /local/home/garzbrya/code/rs/rust2/src/test/ui/async-await/track-caller/panic-track-caller.rs:50:3: 50:15 (#0), segments: [PathSegment { ident: track_caller#0, id: NodeId(179), args: None }], tokens: None }, args: Empty, tokens: None }, tokens: None }), id: AttrId(3), style: Outer, span: /local/home/garzbrya/code/rs/rust2/src/test/ui/async-await/track-caller/panic-track-caller.rs:50:1: 50:16 (#0) }])] }
INFO rustc_ast_lowering::expr self.attrs ⚠️  are SortedMap { data: [] }
```
Which makes me think that we have a separate `LoweringContext` per function in the test? Is that understanding correct? Or is `LoweringContext` encompassing all of the AST?

---

Ideally, there would be some span or ID that represents the entire async fn including the attributes, but I'm not able to find it in the code. At this point in the lowering it seems we are multiple layers removed from the attributes and I haven't found where exactly they are set yet. (Other than being available in `self.attrs`).

Reply:
I found where the `parse_fn` code is so I'm going to trace the logic from there and see if that helps me understand

Reply:
No longer needed (hirId check addressed in master branch already)